graphix/channels.py

```diff
@@ -2,78 +2,137 @@
+    The data should satisfy :math:`\sum K_i^\dagger K_i = I`.
+    """
 
-    Attributes
-    ----------
-    nqubit : int
-        number of qubits acted on by the Kraus operators
-    size : int
-        number of Kraus operators (== Choi rank)
-    kraus_ops : list(dict())
-        the data in format
-        list(dict): [{coef: scalar, operator: list}, {coef: scalar, operator: list}, ...]
+    __nqubit: int
+    __data: list[KrausData]
 
-    """
+    @staticmethod
+    def _nqubit(kraus_data: Iterable[KrausData]) -> int:
+        try:
+            nqubits = set(data.nqubit for data in kraus_data)
+        except ValueError as e:
+            raise ValueError("Failed to intialize KrausChannel object.") from e
+
+        if len(nqubits) != 1:
+            raise ValueError("All operators must have the same shape.")
+        (nqubit,) = nqubits
 
-    def __init__(self, kraus_data):
+        return nqubit
+
+    def __init__(self, kraus_data: Iterable[KrausData]) -> None:
         """Initialize `KrausChannel` given a Kraus operator.
 
         Parameters
         ----------
-        kraus_data : list
-            array of Kraus operator data.
-            list(dict): [{coef: scalar, operator: list}, {parameter: scalar, operator: list}, ...]
-            only works for square Kraus operators
+        kraus_data : Iterable[KrausData]
+            Iterable of Kraus operator data.
 
         Raises
         ------
         ValueError
-            If empty list is provided.
+            If kraus_data is empty.
         """
-        # check there is data
         if not kraus_data:
             raise ValueError("Cannot instantiate the channel with empty data.")
 
-        if not isinstance(kraus_data, (list, np.ndarray, tuple)):
-            raise TypeError(f"The data must be a list, a numpy.ndarray or a tuple not a {type(kraus_data)}.")
+        self.__nqubit = self._nqubit(kraus_data)
+        self.__data = list(copy.deepcopy(data) for data in kraus_data)
 
-        # check that data is correctly formatted before assigning it to the object.
-        assert check_data_values_type(kraus_data)
-        assert check_data_dims(kraus_data)
+        if len(self.__data) > 4**self.__nqubit:
+            raise ValueError("len(kraus_data) cannot exceed 4**nqubit.")
 
-        # check that the channel is properly normalized i.e
-        # \sum_K_i^\dagger K_i = Identity
-        assert check_data_normalization(kraus_data)
-        assert check_rank(kraus_data)
+        # Check that the channel is properly normalized, i.e., \sum_K_i^\dagger K_i = Identity.
+        data = next(iter(self.__data))
+        work = np.zeros_like(data.operator, dtype=np.complex128)
+        for data in self.__data:
+            m = data.coef * data.operator
+            work += m.conj().T @ m
+        if not np.allclose(work, np.eye(2**self.__nqubit)):
+            raise ValueError("The specified channel is not normalized.")
```

Comment:
This code could be vectorize as follows.
```python

        # Check that the channel is properly normalized, i.e., \sum_K_i^\dagger K_i = Identity.
        coefs = np.asarray([data.coef for data in self.__data])
        ops = np.asarray([data.operator for data in self.__data])
        op_sum = np.einsum("i,i...jk,i,i...kl->...jl", coefs, np.transpose(ops.conj(), axes=[0, *range(ops.ndim - 1, 0, -1)]), coefs, ops)
        if not np.allclose(op_sum, np.eye(2**self.__nqubit)):
            raise ValueError("The specified channel is not normalized.")
```

Reply:
I prefer simpler codes here as `KrausChannel.__init__` is very unlikely to be frequently called.